### DISCUSSION TITLE
C++-20  requirement

### DISCUSSION THREAD

**Discussion by m-holger** [qpdf/qpdf-dev#17] [General]
Currently the requirement for qpdf  is stated as:

> A C++ compiler that supports C++-17

For C++-20 we may need to be a bit more specific. Looking for example at the `<format>` header, this is not included in the standard gcc glibc on ubuntu 22.04 based systems.

Looking at support for C++-17 at https://en.cppreference.com/w/cpp/compiler_support/17 

vs C++-20 at https://en.cppreference.com/w/cpp/compiler_support#cpp20

and https://en.cppreference.com/w/cpp/compiler_support#C.2B.2B20_library_features

I think we need to be very selective about which C++-20 feature we use and therefore need to be supported by a compiler in order to build qpdf. The alternative is to postpone C++-20 until qpdf 13.

**Comment by m-holger** [comment]
From https://gcc.gnu.org/onlinedocs/libstdc++/manual/status.html#status.iso.2020 :

> Important: Because the ISO C++20 standard is very recent, GCC's support is experimental.

This makes me lean towards postponement.

**Reply by jberkenbilt** [reply]
Hmm. Probably we should postpone. I thought C++20 was the default for gcc. When it becomes the default, that would be a good signal for us. Alternatively, we could update the c++11 tests to add new language features we depend on, but I don't want qpdf to be bleeding edge -- its place is to be stable and work on as many platforms as possible. I guess I thought c++20 support was better than that.

**Comment by m-holger** [comment]
I think we are in agreement - let's decide to stick with C++-17 for qpdf 12 unless we find a compelling reason to reconsider.

**Comment by m-holger** [comment]
> I think I'd like to go ahead and do the rest of the PointerHolder transition and bump to C++-20. In spite of what the gcc docs say, there are plenty of open source libraries that are standardized on C++-20. I think if we use features that work with gcc, clang, and MSVC, we will be fine. 

To do this, I think we need to be clear which versions we are talking about. For gcc I don't think anything that require > 11 / glibc 2.35 and for LLVM  > 14 makes sense,  given that this is the default for Ubuntu 22.04 . I don't use Windows or Apple, so I do not have a clue what a sensible minimum version to support would be.

Are there any implications for the AppImage build - this  seems to use gcc 7, which I don't think supports C++20.?
